cloud/azure: add file credentials to implicit auth path for storage and KMS

rhu713 opened the pull request:
Augment the credentials lookup chain for the implicit authentication method in
Azure storage and KMS to first look for credentials in a file before proceeding
on to the rest of the chain specified in the DefaultAzureCredential class. The
path of this file is specified by the enviroment variable
`COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE` and, if not empty, is expected
point to a file that provides the Azure Tenant ID, Client ID, and Client Secret
that will be used to access the storage or KMS resource in the YAML format:

```
azure_tenant_id: MY_TENANT_ID
azure_client_id: MY_CLIENT_ID
azure_client_secret: MY_CLIENT_SECRET
```

Epic: [CRDB-28708](https://cockroachlabs.atlassian.net/browse/CRDB-28708)

Release note (enterprise change): Augment the credentials lookup chain for the
implicit authentication method in Azure storage and KMS to first look for
credentials in a file before proceeding on to the rest of the chain specified
in the DefaultAzureCredential class (env vars, Managed Identity, etc.). The
path of this file is specified by the enviroment variable
`COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE` and, if not empty, is expected
point to a file that provides the Azure Tenant ID, Client ID, and Client Secret
that will be used to access the storage or KMS resource in the YAML format:

```
azure_tenant_id: MY_TENANT_ID,
azure_client_id: MY_CLIENT_ID,
azure_client_secret: MY_CLIENT_SECRET
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105647)
<!-- Reviewable:end -->

Commit:
cloud/azure: add file credentials to implicit auth path for storage and KMS

Augment the credentials lookup chain for the implicit authentication method in
Azure storage and KMS to first look for credentials in a file before proceeding
on to the rest of the chain specified in the DefaultAzureCredential class. The
path of this file is specified by the enviroment variable
`COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE` and, if not empty, is expected
point to a file that provides the Azure Tenant ID, Client ID, and Client Secret
that will be used to access the storage or KMS resource in the YAML format:

```
azure_tenant_id: MY_TENANT_ID
azure_client_id: MY_CLIENT_ID
azure_client_secret: MY_CLIENT_SECRET
```

Epic: CRDB-28708

Release note (enterprise change): Augment the credentials lookup chain for the
implicit authentication method in Azure storage and KMS to first look for
credentials in a file before proceeding on to the rest of the chain specified
in the DefaultAzureCredential class (env vars, Managed Identity, etc.). The
path of this file is specified by the enviroment variable
`COCKROACH_AZURE_APPLICATION_CREDENTIALS_FILE` and, if not empty, is expected
point to a file that provides the Azure Tenant ID, Client ID, and Client Secret
that will be used to access the storage or KMS resource in the YAML format:

```
azure_tenant_id: MY_TENANT_ID,
azure_client_id: MY_CLIENT_ID,
azure_client_secret: MY_CLIENT_SECRET
```

adityamaru reviewed (approved):
Everything looks good to me, the tests are really nice!

I would feel even better if you could spin up a roachprod cluster if you haven't already and smoke test:
- Encrypted backup/restore
- Non-encrypted backup/restore

With and without file implicit auth, to ensure there is no regression. Considering we also want to to backport this to 23.1 this might be a good exercise.

rhu713 commented:
bors r+
